storageccl: tolerate rewriting keys with fewer columns than expected

dt opened the pull request:
When rewriting a key, if it belongs to an index in which another is interleaved, once we rewrite past the table/index ID prefix, we previously would assume we had to skip exactly the number of columns indexed before checking if there was an interleave child ID to rewrite.

However, if a span's start or end key does not include all the index columns, this assumption is invalid. It is unclear why we'd have such a span boundary, but if we do, the key rewriter has technically done its job if it rewrites all the IDs in the prefix of a key and then finds the key ends in the middle of the column values. While unexpected, strictly from the point of view of the rewriter, the key's IDs are rewritten and it can return.

Release note (enterprise change): Fix a bug in RESTORE where some unusual range boundaries in interleaved tables caused an error.

dt committed:
storageccl: tolerate rewriting keys with fewer columns than expected

When rewriting a key, if it belongs to an index in which another is interleaved, once we rewrite past the table/index ID prefix, we previously would assume we had to skip exactly the number of columns indexed before checking if there was an interleave child ID to rewrite.

However, if a span's start or end key does not include all the index columns, this assumption is invalid. It is unclear why we'd have such a span boundary, but if we do, the key rewriter has technically done its job if it rewrites all the IDs in the prefix of a key and then finds the key ends in the middle of the column values. While unexpected, strictly from the point of view of the rewriter, the key's IDs are rewritten and it can return.

Release note (enterprise change): Fix a bug in RESTORE where some unusual range boundaries in interleaved tables caused an error.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35750)
<!-- Reviewable:end -->

dt commented:
I'm still not sure how we got the strange span that led to this (we're going back to the customer's cluster to look at the actual ranges), so I'm not quite sure how to test this, but does this seem completely crazy?

madelynnblue reviewed (approved):
Uh weird. The code LGTM although I'm fairly scared to merge it since we have neither a test nor even a confirmation from a user that it works.

dt commented:
I confirmed this fixes RESTORE’s handling of  the offending backup we were provided.

That backup quite obviously has a span with a shorter than expected key — but i’m still unsure what led to such a span in a backup in the first place. But given that we have evidence such backups exist — since we have one — I figure we should make restore handle them either way.

dt commented:
I'm planning on sitting on this until we hear back about the actual ranges on the cluster that produced the backup in question, in case we can determine where such a range boundary came from, since that might also inform how to test it.

dt commented:
So we heard back from the customer and that short key is in their existing range boundary, so this isn’t something that BACKUP did / made up.

I’m still not sure how to get that split though.

tim-o commented:
Zendesk ticket #[3190](https://cockroachdb.zendesk.com/agent/tickets/3190) has been linked to this issue.

dt commented:
took another look at this this morning and realized that split comes from partitioning, then realized why the split I had added to the test without success didn't quite do the same thing. Updated the test with a repro so I think this is good to go now.

madelynnblue reviewed (approved):
Oh good find.

dt commented:
bors r+

dt commented:
this will need 19.1 and 2.1 backports
